### DISCUSSION TITLE
Sorting includes

### DISCUSSION THREAD

**Discussion by henryiii** [pypa/cibuildwheel#518] [Ideas]
We've already been dealing with occasion merge conflicts because of changes on imports (glaring at typing right now...). There are two solutions: One is https://pypi.org/project/reorder-python-imports/, via pre-commit. That even does the smart grouping following PEP8, but it does produces this rewrite:

```python
from typing import Any
from typing import Dict
from typing import List
from typing import Tuple
```

See [reasoning here](https://github.com/asottile/reorder_python_imports#why-this-style). Maybe I could get used to it; part of the problem of having too many characters is solved by the smart grouping, I'm more used to the following, black compatible import style:

```python
from typing import (
    Any,
    Dict,
    List,
    Tuple,
)
```

This does have a slightly higher chance of merge conflicts, but still solves most of them without adding a massive number of extra characters and visual clutter (it does add two lines, though). Generally I use black + isort + a flake8 grouping check to get this result; we could probably get most of the way there with isort + flake8, though, maybe with a pre-run of reorder-python-imports then manual grouping?

Note that `reorder_python_imports` has some useful tools unrelated to reordering, like removing/adding imports.

**Comment by Czaki** [comment]
Why not isort?

**Reply by henryiii** [reply]
Isn't that what I listed as the second solution? A flake8 check + isort. isort can't get the grouping for you like reorder-python-imports can.

Actually I think I forgot to list it as the second solution. I intended to when I mentioned it.

**Reply by henryiii** [reply]
Actually, I think I've only done the flake8 + black before, isort is quite powerful, didn't realize how much so. Nice!

**Reply by henryiii** [reply]
Made a PR for people to look at.

**Comment by mayeut** [comment]
If adding `isort`, I'd really be in favor of adding `black` as well. I think given the current status of `cibuildwheel`,  having `mypy`, `flake8`, `black` & `isort` would be great.

**Reply by joerick** [reply]
I did try `black` a while back on the cibuildwheel source, but I didn't like the result. We do use a lot of long lines, but if you split them up like `black` does, it becomes very hard to follow the flow of the program. (Also, it made me realise how much whitespace is cultural, and often the programmer lays something out with certain whitespace because it's easier to understand, IMO.)

**Reply by Czaki** [reply]
`black` allow configuring the length of lines. I set the line length to 120.
https://github.com/4DNucleome/PartSeg/blob/7d386edd0fc00ec98caedda5ecce0e605665f1ef/pyproject.toml#L10

**Reply by henryiii** [reply]
Black isn't just a fire and forget do-nothing tool - you learn to structure your code to format nicely. I think it improves the overall code as well. You avoid dumping too much on one line, you use more temporaries instead of nesting things, and it makes several structures (like nested for loops in generator expressions) actually usable.

Once  &nbsp; you &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; have &nbsp; &nbsp; gotten used &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; to &nbsp; &nbsp; &nbsp; reading &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Blacked &nbsp; code, you'll &nbsp; &nbsp; start to &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; read code &nbsp; faster, &nbsp; &nbsp; &nbsp; &nbsp; since &nbsp; &nbsp; &nbsp; developer &nbsp; &nbsp; preferences &nbsp; in whitespace &nbsp; &nbsp; &nbsp; add &nbsp; &nbsp; &nbsp; &nbsp; an extra&nbsp; &nbsp; layer &nbsp; you have to &nbsp; parse &nbsp; &nbsp; &nbsp; visually.

Once you have set standards, you read faster, because you no longer have to slow down and carefully see if a developer wrote `{[` somewhere because they were "grouping" this in their head. There's a reason we don't allow random whitespace in written human communication, why allow it in code?

And, if the whitespace is special, _just turn off black for that special block_. Clearly tells the user formatting is important, too.

Also, Black is designed to minimize diffs, which is a good design goal. :)

I think the longer lines would be useful for cibuildwheel, because it's designed with pretty deep indentation in some places, and long identifiers.

**Reply by henryiii** [reply]
Eventually I'll probably take a stab at running Black, just to see what I can get. If it looks reasonable, I'll show it. If not, I'll hide it away and pretend it never happened. ;P

**Reply by YannickJadoul** [reply]
> Once   you           have     gotten used           to       reading               Blacked   code, you'll     start to           read code   faster,         since       developer     preferences   in whitespace       add         an extra    layer   you have to   parse       visually.

That's not very useful or meaningul whitespace, though, so a bit of a false argument ;-)

**Reply by YannickJadoul** [reply]
Also

> There's a reason we don't allow random whitespace in written human communication, why allow it in code?

We do. Not after periods or commas. But we do allow newlines vs. stuffing something at the end of a sentence if the core of the sentence is clear.
Also, e.g. "also" doesn't always take a comma, and depends on stress and emphasis. And this kind of extra layer (stress/emphasis) is exactly what any rigid code formatter removes. It should _never_ be critical (just like you can read this sentence without the italics), but it can help nudging you into the correct direction (imo).


> And, if the whitespace is special, _just turn off black for that special block_.

But I agree it should (almost) never be important enough to clutter up the code and distract with extra comments that turn it off.

**Reply by YannickJadoul** [reply]
(I've most often seen comments about some style being confusing attended to, btw. So if there's some grouping or whitespace or... that works for someone, but is confusing, review could/should catch it as well?)

**Reply by henryiii** [reply]
Black lets you add an empty line too,
to allow grouping,
just like we do after sentences.
It also lets you add an extra comma at the end of a list to signify something.

**Reply by henryiii** [reply]
Unblacked:

```python
x = [[[0.0, 1.1, 2.2, 3.3],
      []],
      [[4.4, 5.5]],
      [],
      [[6.6],
      [7.7, 8.8, 9.9, 10.0, 11., 12.]]]
```

Blacked:
```python
x = [
    [[0.0, 1.1, 2.2, 3.3], []],
    [[4.4, 5.5]],
    [],
    [[6.6], [7.7, 8.8, 9.9, 10.0, 11.0, 12.0]],
]
```

If I'm reading unBlacked code, I have to be wary of the first structure - it's legal, after all. If I'm reading Blacked code, that can never happen, I no longer pause to count the brackets, and can just keep going. Even if no one was this evil/clever, I still have to check.

If you need to impart meaning to the reader, you should impart it to Python too.

**Reply by YannickJadoul** [reply]
> ```python
> x = [[[0.0, 1.1, 2.2, 3.3],
>       []],
>      [[4.4, 5.5]],
>      [],
>      [[6.6],
>       [7.7, 8.8, 9.9, 10.0, 11., 12.]]]
> ```

Two spaces less already paints a difference picture (one that Black doesn't perfectly show either, I'd say, but that's disputable). But I've never said that _anything_ goes. The first one is more or less objectively badly formatted (especially without meaning attached); I think we can agree on that.

The thing I don't agree on is that in every single case, Black's output is best. Maybe, in this specific case, maybe the third level is the one you want/need to focus on? Maybe all of those deserve their own line, and meanwhile the structure is also clear from context?

> If you need to impart meaning to the reader, you should impart it to Python too.

Either I don't understand what you mean, or I don't agree on that. By that reasoning, comments and well-chosen names don't matter either, because Python doesn't attach meaning to these either?

**Reply by henryiii** [reply]
I don't claim black is perfect, or the best output. I claim it is a standard (and minimizes diffs). You will never have a perfect standard.

> Maybe, in this specific case, maybe the third level is the one you want/need to focus on? 

But you are tricking your reader, but not Python. If a reader can be sure formatting is standard, they don't have to check for it. Your structures should be clear to the reader. If there's a format on/off, then they will know to slow down and check each bracket; but most of the time, are you trying to hide nesting to the reader? I've seen users do it, certainly, but I've never found it helpful. I want to know what the _actual_ structure is.

That's what yaml is good at, actually, having clear structure humans can parse quickly.

> By that reasoning, comments and well-chosen names don't matter either

Not really, comments and names are explicitly how you communicate to the reader. The structure of your program, however, is something you are communicating to Python (blocks, brackets, etc); they do not have arbitrary meaning.

**Reply by YannickJadoul** [reply]
> But you are tricking your reader, but not Python.

I'm not trying to trick anyone. I'm trying to put emphasis on what matters to the reader.

> > By that reasoning, comments and well-chosen names don't matter either
> 
> Not really, comments and names are explicitly how you communicate to the reader. The structure of your program, however, is something you are communicating to Python (blocks, brackets, etc); they do not have arbitrary meaning.

That's not true to me. Python's syntactic structure is not fixed, and allows you to communicate through that as well. That's exactly the small gap where you can (sometimes) express your goal/focus, next to satisfying Python's syntax. Just like Python tells you to make variables, but not what name they should have, so the wiggle room allows you to communicate to the user. In any case, as long as it doesn't confuse the reader, it can only help.
